### PR TITLE
fix: Use unicode_segmentation to truncate INSERT statement

### LINKED ISSUE #8010: bug: query server randomly panics on INSERT statements containing utf8 data
### Search before asking

- [X] I had searched in the [issues](https://github.com/datafuselabs/databend/issues) and found no similar issues.

### Version

v0.8.45-nightly-59c3b86-simd(1.66.0-nightly-2022-09-23T09:02:25.327433275Z)

### What's Wrong?

When inserting data that contains multi bytes UTF-8 characters, query server randomly panics.

It happens when the `INSERT` statement is bigger than 64 bytes, then when filling the query context, the `INSERT` statement is truncated to 64 byte by the `common_legacy_parser::sql_common::SQLCommon::short_sql` function.

Unfortunately, if a multibyte character is located around the 64 position, it panics. This is happening in the following function:

https://github.com/datafuselabs/databend/blob/b8f1a2979b3bd957faac0c9fc77dc3b61fe47fe8/src/query/legacy-parser/src/sql_common.rs#L126-L133

Some trace from the `databend-query` server logs:

```
panicked at 'byte index 64 is not a char boundary; it is inside '災' (bytes 63..66) of `INSERT INTO `post` VALUES (781725994,'2012/2/19 東日本大震災復興ＮＰＯ支援・全国プロ
ジェクト第２回支援フォーラム','<p>2011/4/24 立正大学会場でのフォーラムで設立した東日本大震災復興ＮＰＯ支援・`[...]', src/query/legacy-parser/src/sql_common.rs:129:31"
```

Partial backtrace:
```
Backtrace [{ fn: "common_tracing::panic_hook::log_panic", file: "/workspace/src/common/tracing/src/panic_hook.rs", line: 33 },
 { fn: "std::panicking::rust_panic_with_hook", file: "/rustc/98ad6a5519651af36e246c0335c964dd52c554ba/library/std/src/panicking.rs", line: 702 },
 { fn: "std::panicking::begin_panic_handler::{{closure}}", file: "/rustc/98ad6a5519651af36e246c0335c964dd52c554ba/library/std/src/panicking.rs", line: 588 },
 { fn: "std::sys_common::backtrace::__rust_end_short_backtrace", file: "/rustc/98ad6a5519651af36e246c0335c964dd52c554ba/library/std/src/sys_common/backtrace.rs", line: 138 },
 { fn: "rust_begin_unwind", file: "/rustc/98ad6a5519651af36e246c0335c964dd52c554ba/library/std/src/panicking.rs", line: 584 },
 { fn: "core::panicking::panic_fmt", file: "/rustc/98ad6a5519651af36e246c0335c964dd52c554ba/library/core/src/panicking.rs", line: 142 },
 { fn: "core::str::slice_error_fail_rt", file: "/rustc/98ad6a5519651af36e246c0335c964dd52c554ba/library/core/src/str/mod.rs" },
 { fn: "core::str::slice_error_fail", file: "/rustc/98ad6a5519651af36e246c0335c964dd52c554ba/library/core/src/str/mod.rs", line: 86 },
 { fn: "common_legacy_parser::sql_common::SQLCommon::short_sql", file: "/workspace/src/query/legacy-parser/src/sql_common.rs" },
 { fn: "databend_query::sessions::query_ctx_shared::QueryContextShared::attach_query_str", file: "/workspace/src/query/service/src/sessions/query_ctx_shared.rs", line: 243 },
 { fn: "<databend_query::sessions::query_ctx::QueryContext as common_catalog::table_context::TableContext>::attach_query_str", file: "/workspace/src/query/service/src/sessions/query_ctx.rs", line: 251 },
 { fn: "databend_query::servers::http::v1::query::execute_state::ExecuteState::try_start_query::{{closure}}", file: "/workspace/src/query/service/src/servers/http/v1/query/execute_state.rs", line: 227 },
 { fn: "<core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll", file: "/rustc/98ad6a5519651af36e246c0335c964dd52c554ba/library/core/src/future/mod.rs", line: 91 },
 { fn: "databend_query::servers::http::v1::query::http_query::HttpQuery::try_create::{{closure}}::{{closure}}", file: "/workspace/src/query/service/src/servers/http/v1/query/http_query.rs", line: 249 },
```

### How to Reproduce?

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by zenria**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Use unicode_segmentation crate to extract graphemes to truncate big `INSERT` statements instead of directly indexing first 64 bytes of the statement, which is unsafe: it panics if a multibyte unicode character is spread between the 63th & 64th byte of the string. 

Fixes #8010

### CODE REVIEW

**Comment by sundy-li** [comment]
Good job, need a test case to cover this bug.

File: src/query/legacy-parser/src/sql_common.rs [hunk 1]
```diff
@@ -126,7 +127,17 @@ impl SQLCommon {
     pub fn short_sql(query: &str) -> String {
         let query = query.trim_start();
         if query.len() >= 64 && query[..6].eq_ignore_ascii_case("INSERT") {
-            format!("{}...", &query[..64])
+            // keep first 64 graphemes
```
[the remark this answers is not in the document]
Comment (zenria) [reply src/query/legacy-parser/src/sql_common.rs:130]: `query` is already valid unicode, so I think there is no need to use `bstr` `chars`.

I do not think using the `graphemes` iterator have a significant performance impact over`bstr` `chars` , in all case, a new `String` must be allocated to contain the truncated statement.

**Review by sundy-li — APPROVED** [review]

**Review by PsiACE — APPROVED** [review]

**Comment by Xuanwo** [comment]
@zenria Hi, we use mergify to automate the workflows. Please [authorize modification on its head branch](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) so that we can update your PRs.

**Comment by zenria** [comment]
@Xuanwo Ok, but I can't find the `Allow edits from maintainers` checkbox: the right column of the PR does not have this.

It seems to appear only "On user-owned forks", but I forked the repo using an organisation...

**Comment by Xuanwo** [comment]
@mergify update

**Comment by Xuanwo** [comment]
> It seems to appear only "On user-owned forks", but I forked the repo using an organisation...

Well, we need to merge the `main` branch by hand now.

**Comment by Xuanwo** [comment]
All tests have passed and not other changes, let's merge.

**Comment by Xuanwo** [comment]
@zenria Thanks for your contribution!

**Pull request merged**